sql: TestIdleCancelSession failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/cfd53f62f038a06e34570f11ec02bc74dabd2aab

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestIdleCancelSession PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1021766&tab=buildLog

```
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1249 +0x1364
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:680 +0xac

Goroutine 589 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x56d
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:499 +0x766
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:249 +0x141
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Query()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:233 +0x10e
  github.com/cockroachdb/cockroach/pkg/server.(*Server).clusterVersion()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:162 +0xcd
  github.com/cockroachdb/cockroach/pkg/server.(*Server).upgradeStatus()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:106 +0x8f
  github.com/cockroachdb/cockroach/pkg/server.(*Server).startAttemptUpgrade.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server_update.go:64 +0x27b
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0xf3

Goroutine 1273 (running) created at:
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xb8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:417 +0x1446
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:457 +0x87a
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:676 +0x16e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:631 +0x53
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:639 +0x34
==================
I181121 08:04:55.513024 89809 server/server_update.go:68  [n2] no need to upgrade, cluster already at the newest version
I181121 08:04:55.524198 90052 sql/event_log.go:126  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:36951} Attrs: Locality: ServerVersion:2.1-3 BuildTag:v2.2.0-alpha.20181119-229-gcfd53f6 StartedAt:1542787495052461629 LocalityAddress:[]} ClusterID:b9af0aae-d94e-4be7-b475-15ad76827f16 StartedAt:1542787495052461629 LastUp:1542787495052461629}
I181121 08:04:55.553948 90105 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181121 08:04:55.554007 90106 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181121 08:04:55.587560 90105 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
I181121 08:04:55.656168 89232 kv/transport_race.go:113  transport race promotion: ran 22 iterations on up to 301 requests
--- FAIL: TestIdleCancelSession (2.36s)
	testing.go:730: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/7efc92a4dec689efc855ecd382a6f6b6065b98ec

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestIdleCancelSession PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1055988&tab=buildLog

```
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1011 +0xde7
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1249 +0x1364
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:680 +0xac

Goroutine 837 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x56d
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:489 +0x766
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:239 +0x141
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Query()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:223 +0x10e
  github.com/cockroachdb/cockroach/pkg/sql.(*LeaseManager).DeleteOrphanedLeases.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/lease.go:1786 +0x342
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0x14a

Goroutine 486 (running) created at:
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xb8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:424 +0x1483
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:464 +0x87a
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:676 +0x16e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:631 +0x53
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:639 +0x34
==================
I181214 13:21:53.461014 97125 util/stop/stopper.go:536  quiescing; tasks left:
1      ts.poller: poll
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181214 13:21:53.463185 97124 util/stop/stopper.go:536  quiescing; tasks left:
1      node.Node: batch
1      [async] transport racer
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
W181214 13:21:53.480139 96896 ts/db.go:195  [n2,ts-poll] error writing time series data: failed to send RPC: sending to all 1 replicas failed; last error: <nil> rpc error: code = Unavailable desc = transport is closing
I181214 13:21:53.480592 97125 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181214 13:21:53.481667 96754 kv/transport_race.go:113  transport race promotion: ran 36 iterations on up to 310 requests
I181214 13:21:53.483326 97124 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181214 13:21:53.484182 97125 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
I181214 13:21:53.497581 97124 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
--- FAIL: TestIdleCancelSession (2.22s)
	testing.go:730: race detected during execution of test

```

SHA: https://github.com/cockroachdb/cockroach/commits/8367c883c5db0f4b5aea949530e41a068f25530d

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestIdleCancelSession PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1058444&tab=buildLog

```
  github.com/cockroachdb/cockroach/pkg/sql.glob..func3()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_fsm.go:276 +0x59
  github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:105 +0x20e
  github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:134 +0x165
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1880 +0x1fb
  github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1266 +0x159a
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:203 +0x7d

Goroutine 71 (running) created at:
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:678 +0xb8
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:424 +0x1483
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/transport/http2_server.go:464 +0x87a
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:676 +0x16e
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func2()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:631 +0x53
  github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleRawConn.func3()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:639 +0x34

Goroutine 1048 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:202 +0x56d
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:489 +0x766
  github.com/cockroachdb/cockroach/pkg/sql.(*internalExecutorImpl).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:239 +0x141
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).Query()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:223 +0x10e
  github.com/cockroachdb/cockroach/pkg/sql.(*LeaseManager).DeleteOrphanedLeases.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/lease.go:1786 +0x342
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0x14a
==================
I181217 15:07:37.052285 96786 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181217 15:07:37.055436 96787 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181217 15:07:37.058641 96395 kv/transport_race.go:113  transport race promotion: ran 27 iterations on up to 312 requests
I181217 15:07:37.060173 96786 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
1      [async] closedts-rangefeed-subscriber
I181217 15:07:37.064944 96786 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-subscription
I181217 15:07:37.066022 96787 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
--- FAIL: TestIdleCancelSession (2.66s)
	testing.go:730: race detected during execution of test

```

--- a/pkg/sql/txn_state.go
+++ b/pkg/sql/txn_state.go
@@ -45,9 +45,11 @@ type txnState struct {
 	// Mutable fields accessed from goroutines not synchronized by this txn's
 	// session, such as when a SHOW SESSIONS statement is executed on another
 	// session.
+	//
 	// Note that reads of mu.txn from the session's main goroutine do not require
 	// acquiring a read lock - since only that goroutine will ever write to
-	// mu.txn.
+	// mu.txn. Writes to mu.txn do require a write lock to guarantee safety with
+	// reads by other goroutines.
 	mu struct {
 		syncutil.RWMutex
 
@@ -246,7 +248,9 @@ func (ts *txnState) finishSQLTxn() {
 	ts.sp.Finish()
 	ts.sp = nil
 	ts.Ctx = nil
+	ts.mu.Lock()
 	ts.mu.txn = nil
+	ts.mu.Unlock()
 	ts.recordingThreshold = 0
 }
 
@@ -269,11 +273,16 @@ func (ts *txnState) finishExternalTxn() {
 	}
 	ts.sp = nil
 	ts.Ctx = nil
+	ts.mu.Lock()
 	ts.mu.txn = nil
+	ts.mu.Unlock()
 }
 
 func (ts *txnState) setPriority(userPriority roachpb.UserPriority) error {
-	if err := ts.mu.txn.SetUserPriority(userPriority); err != nil {
+	ts.mu.Lock()
+	err := ts.mu.txn.SetUserPriority(userPriority)
+	ts.mu.Unlock()
+	if err != nil {
 		return err
 	}
 	ts.priority = userPriority

